adam/learner/propose_but_verify.py

```diff
@@ -0,0 +1,372 @@
+
+        rng = Random()
+        rng.seed(params.optional_integer("random_seed", default=0))
+
+        return ProposeButVerifyLanguageLearner(
+            graph_match_confirmation_threshold=params.floating_point(
+                "graph_match_confirmation_threshold"
+            ),
+            graph_logger=graph_logger,
+            log_word_hypotheses_to=log_word_hypotheses_dir,
+            rng=rng,
+            ontology=GAILA_PHASE_1_ONTOLOGY,
+        )
+
+    # observe method
+    def observe(
+        self, learning_example: LearningExample[PerceptionT, LinguisticDescription]
+    ) -> None:
+
+        logging.info("Observation %s", self._observation_num)
+        self._observation_num += 1
+
+        perception = learning_example.perception
+        if len(perception.frames) != 1:
+            raise RuntimeError("Pursuit learner can only handle single frames for now")
+        if isinstance(perception.frames[0], DevelopmentalPrimitivePerceptionFrame):
+            original_perception_graph = PerceptionGraph.from_frame(
+                perception.frames[0]
+            ).copy_as_digraph()
+        else:
+            raise RuntimeError("Cannot process perception type.")
+        # Remove learner from the perception
+        observed_perception_graph = graph_without_learner(original_perception_graph)
+        observed_linguistic_description = (
+            learning_example.linguistic_description.as_token_sequence()
+        )
+
+        self.learn_propose_but_verify(
+            observed_perception_graph, observed_linguistic_description
+        )
+
+    # learn method
+    def learn_propose_but_verify(
+        self,
+        observed_perception_graph: PerceptionGraph,
+        observed_linguistic_description: Tuple[str, ...],
+    ) -> None:
+        logging.info(f"Pursuit learner observing {observed_linguistic_description}")
+        for word in observed_linguistic_description:
+            if word in ("a", "the"):
+                continue
+            logging.info(f"Considering '{word}'")
+            if word not in self._words_to_hypotheses:
+                # a) If word has never been seen before, go through initialization step
+                self.initialization_step(word, observed_perception_graph)
+            else:
+                # b) If word has been seen before, test the recorded hypothesis
+                if self.test_hypothesis_step(word, observed_perception_graph):
+                    # If hypothesis is confirmed, do nothing
+                    continue
+                else:
+                    # If hypothesis is not confirmed, treat word as an unseen word, and go
+                    # through the initialization step again (to randomly select meaning)
+                    self.initialization_step(word, observed_perception_graph)
+            if self._log_word_hypotheses_to:
+                self._log_hypotheses(word)
+
+    def initialization_step(self, word: str, observed_perception_graph: PerceptionGraph):
```

Comment:
Might be better to name it `initialize_new_meaning()` or so, since we no longer have "learning steps" like we used to in pursuit.

Reply:
Renamed. I wasn't sure if a naming convention (i.e: with the word 'step')